### ISSUE TITLE
pylance should not watch a file recursively - python310.zip is not directory

### ISSUE THREAD

**Issue by idimov-keeper** [microsoft/pylance-release#2890]
see also [Watcher should not allow to watch a file recursively or resort to flat watching](https://github.com/microsoft/vscode/issues/151009#)

## Environment data

-   Language Server version: Pylance language server 2022.6.0 (pyright c6e1d8dd)
-   OS and version: OS Name	Microsoft Windows 11 Pro, Version	10.0.22000 Build 22000
-   Python version (& distribution if applicable, e.g. Anaconda): Python 3.10.4

## Code Snippet

```python
print('Hello, World!')
```

## Repro Steps

1.  Python 3.10.4 - latest embeddable version from official website[Windows embeddable package (64-bit)](https://www.python.org/ftp/python/3.10.4/python-3.10.4-embed-amd64.zip)
2.  Add python to %PATH% env var and start vs code
3. Vs Code - create python project
4. Apparently pylance thinks python310.zip is a directory

## Expected behavior

pylance to properly determine if python.zip is a file or directory and to monitor it accordingly.

## Actual behavior

[2022-05-31 17:41:13.382] [renderer1] [error] [File Watcher (parcel)] Unexpected error: Not a directory (EUNKNOWN) (path: c:\PROGRAMS\python\python\python310.zip)
[2022-05-31 17:41:13.387] [renderer1] [error] [File Watcher (universal)] restarting watcher after error: Not a directory

and another warning on non-existent directory (there's only one file in ...\src\ - test.py that print hello world)
[Warn  - 12:13:37 PM] (25888) stubPath d:\WORK\tmp\empty.py\src\typings is not a valid directory.

## Logs - truncated due to 65K limit error: body is too long (maximum is 65536 characters)
see attachment for full log trace

```
[Info  - 12:27:13 PM] (2000) Pylance language server 2022.6.0 (pyright c6e1d8dd) starting
[Info  - 12:27:13 PM] (2000) Server root directory: D:\PROGRAMS\VSCode\data\extensions\ms-python.vscode-pylance-2022.6.0\dist
Notebook support: Legacy
[Info  - 12:27:13 PM] (2000) No configuration file found.
[Info  - 12:27:13 PM] (2000) No pyproject.toml file found.
[Info  - 12:27:13 PM] (2000) Setting pythonPath for service "src": "C:\PROGRAMS\python\python\python.exe"
[Warn  - 12:27:13 PM] (2000) stubPath d:\WORK\tmp\empty.py\src\typings is not a valid directory.
[Info  - 12:27:14 PM] (2000) Assuming Python version 3.10
[Info  - 12:27:14 PM] (2000) Assuming Python platform Windows
[Info  - 12:27:14 PM] (2000) Search paths for d:\WORK\tmp\empty.py\src
[Info  - 12:27:14 PM] (2000)   D:\PROGRAMS\VSCode\data\extensions\ms-python.vscode-pylance-2022.6.0\dist\typeshed-fallback\stdlib
[Info  - 12:27:14 PM] (2000)   d:\WORK\tmp\empty.py\src
[Info  - 12:27:14 PM] (2000)   d:\WORK\tmp\empty.py\src\typings
[Info  - 12:27:14 PM] (2000)   D:\PROGRAMS\VSCode\data\extensions\ms-python.vscode-pylance-2022.6.0\dist\typeshed-fallback\stubs\...
[Info  - 12:27:14 PM] (2000)   D:\PROGRAMS\VSCode\data\extensions\ms-python.vscode-pylance-2022.6.0\dist\bundled\stubs
[Info  - 12:27:14 PM] (2000)   c:\PROGRAMS\python\python3.10.4\python310.zip
[Info  - 12:27:14 PM] (2000)   c:\PROGRAMS\python\python3.10.4
[Info  - 12:27:14 PM] (2000)   c:\PROGRAMS\python\python3.10.4\Lib\site-packages
[Info  - 12:27:14 PM] (2000)   c:\PROGRAMS\python\python3.10.4\Lib\site-packages\win32
[Info  - 12:27:14 PM] (2000)   c:\PROGRAMS\python\python3.10.4\Lib\site-packages\win32\lib
[Info  - 12:27:14 PM] (2000)   c:\PROGRAMS\python\python3.10.4\Lib\site-packages\pythonwin
[Info  - 12:27:14 PM] (2000) Adding fs watcher for library directories:
 C:\PROGRAMS\python\python\python310.zip
C:\PROGRAMS\python\python
C:\PROGRAMS\python\python\lib\site-packages
C:\PROGRAMS\python\python\lib\site-packages\win32
C:\PROGRAMS\python\python\lib\site-packages\win32\lib
C:\PROGRAMS\python\python\lib\site-packages\Pythonwin
[Info  - 12:27:14 PM] (2000) Adding fs watcher for directories:
 d:\WORK\tmp\empty.py\src
[Info  - 12:27:14 PM] (2000) Searching for source files
[Info  - 12:27:14 PM] (2000) Found 1 source file
(2000) [FG] parsing: d:\WORK\tmp\empty.py\src\test.py (16ms)
(2000) [FG] parsing: D:\PROGRAMS\VSCode\data\extensions\ms-python.vscode-pylance-2022.6.0\dist\typeshed-fallback\stdlib\builtins.pyi [fs read 3ms] (89ms)
(2000) [FG] binding: D:\PROGRAMS\VSCode\data\extensions\ms-python.vscode-pylance-2022.6.0\dist\typeshed-fallback\stdlib\builtins.pyi (27ms)
(2000) [FG] binding: d:\WORK\tmp\empty.py\src\test.py (0ms)
[Info  - 12:27:14 PM] (2000) Background analysis(1) root directory: D:\PROGRAMS\VSCode\data\extensions\ms-python.vscode-pylance-2022.6.0\dist
[Info  - 12:27:14 PM] (2000) Background analysis(1) started
(2000) Background analysis message: setConfigOptions
(2000) Background analysis message: setImportResolver
(2000) Background analysis message: ensurePartialStubPackages
(2000) Background analysis message: setTrackedFiles
(2000) Background analysis message: markAllFilesDirty
(2000) Background analysis message: setFileOpened
(2000) Background analysis message: getDiagnosticsForRange
(2000) Background analysis message: analyze
(2000) [BG(1)] analyzing: d:\WORK\tmp\empty.py\src\test.py ...
(2000) [BG(1)]   parsing: d:\WORK\tmp\empty.py\src\test.py (21ms)
(2000) [BG(1)]   parsing: D:\PROGRAMS\VSCode\data\extensions\ms-python.vscode-pylance-2022.6.0\dist\typeshed-fallback\stdlib\builtins.pyi [fs read 3ms] (122ms)
(2000) [BG(1)]   binding: D:\PROGRAMS\VSCode\data\extensions\ms-python.vscode-pylance-2022.6.0\dist\typeshed-fallback\stdlib\builtins.pyi (37ms)
(2000) [BG(1)]   binding: d:\WORK\tmp\empty.py\src\test.py (0ms)
(2000) [BG(1)]   checking: d:\WORK\tmp\empty.py\src\test.py ...
(2000) [BG(1)]     parsing: D:\PROGRAMS\VSCode\data\extensions\ms-python.vscode-pylance-2022.6.0\dist\typeshed-fallback\stdlib\typing_extensions.pyi [fs read 7ms] (19ms)
(2000) [BG(1)]     binding: D:\PROGRAMS\VSCode\data\extensions\ms-python.vscode-pylance-2022.6.0\dist\typeshed-fallback\stdlib\typing_extensions.pyi (2ms)
(2000) [BG(1)]     parsing: D:\PROGRAMS\VSCode\data\extensions\ms-python.vscode-pylance-2022.6.0\dist\typeshed-fallback\stdlib\typing.pyi [fs read 1ms] (37ms)
(2000) [BG(1)]     binding: D:\PROGRAMS\VSCode\data\extensions\ms-python.vscode-pylance-2022.6.0\dist\typeshed-fallback\stdlib\typing.pyi (12ms)
[Info  - 12:27:14 PM] (2000) Indexer background runner(2) root directory: 

...

(2000) [IDX(2)]   index execution environment d:\WORK\tmp\empty.py\src [found 10661 in 479 files] (1154ms)
(2000) [IDX(2)] index libraries d:\WORK\tmp\empty.py\src [found 10661 in 1 exec envs] (1893ms)
[Info  - 12:27:16 PM] (2000) Indexer done(2). indexed 479 files

```
[pylance.log](https://github.com/microsoft/pylance-release/files/8840444/pylance.log)

**Comment by erictraut** [comment]
Are you seeing any user-visible behaviors (other than the error in the log) related to this issue? Or are you just concerned about the log? The log is intended for debugging issues when pylance is behaving in unintended or undesirable ways, so if you're not seeing any such issues, you can ignore the error in the log.

**Comment by bpasero** [comment]
See https://github.com/microsoft/vscode/issues/151009#

**Comment by erictraut** [comment]
I still can't tell from that report whether there's an undesired user-visible behavior or simply a concern about a line seen in a debugging log.

Pylance does have a virtual file system (VFS) that "mounts" zip files and reads Python source files from them. This is needed to support zip-based "eggs" which was a popular way to distribute packages a long time ago. I suspect that the VFS is incorrectly attempting to install a file watcher on the virtual directory. But as I mentioned, I'm still not clear on whether there's a real problem here.

**Comment by bpasero** [comment]
As far as I can tell, the entirety of file watching breaks as a consequence because Parcel watcher throws an error, we try to restart 5 times and then give up.

**Comment by idimov-keeper** [comment]
This happens and reloads don't help
![image](https://user-images.githubusercontent.com/78815270/172189546-8f4145a8-1a9f-4bbd-aa0e-76101cba0ad7.png)

**Comment by heejaechang** [comment]
it might be because we are giving zip as folder to file watcher since zip is returned as import roots.

**Comment by debonte** [comment]
This issue has been fixed in version 2022.6.10, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2022610-8-June-2022)
